### PR TITLE
config: simplify default config handling.

### PR DESCRIPTION
**Pull request by fabxc**
As suggested by @neelance, this makes default configuration handling more transparent.
Thanks again @neelance, I like it.

### CODE REVIEW

**Comment by neelance** [comment]
Glad I could help. :-)

File: config/config.go [hunk 1]
```diff
@@ -172,27 +152,18 @@ type DefaultedGlobalConfig struct {
...
-}
-
 // UnmarshalYAML implements the yaml.Unmarshaler interface.
-func (c *ScrapeConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
-	c.DefaultedScrapeConfig = DefaultScrapeConfig
-	err := unmarshal(&c.DefaultedScrapeConfig)
-	if err != nil {
+func (c *GlobalConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultGlobalConfig
+	type plain GlobalConfig
```
Comment (beorn7) [inline config/config.go]: Please add a comment with a hint why we do this. ;)

Comment (fabxc) [inline config/config.go]: Comment added on first occurrence.

File: config/config.go [hunk 2]
```diff
@@ -45,19 +45,19 @@ func LoadFromFile(filename string) (*Config, error) {
...
 
 	// The default global configuration.
-	DefaultGlobalConfig = DefaultedGlobalConfig{
+	DefaultGlobalConfig = GlobalConfig{
 		ScrapeInterval:     Duration(1 * time.Minute),
 		ScrapeTimeout:      Duration(10 * time.Second),
 		EvaluationInterval: Duration(1 * time.Minute),
 	}
 
 	// Te default scrape configuration.
```
Comment (beorn7) [inline config/config.go]: Typo te -> the

**Comment by beorn7** [comment]
Dumb question: I don't fully understand how the yaml marshaling works under the hood, but it looks to me that the default config structs are actually partially overwritten, thus modifying the defaults for future config loads. The test just loads once and cannot prove that this is not happening.

Perhaps discuss in person tomorrow...

**Comment by fabxc** [comment]
We copy the values of the default structs into the configuration structs. Then we unmarshal into the latter.
So inside of the package the default structs are only written on definition.

We can take a closer look tomorrow to be sure.

**Comment by fabxc** [comment]
So yep, good catch. It was indeed a problem for the `GlobalConfig` pointer in `DefaultConfig`. Test added and fixed.

File: config/testdata/glob_timeout.good.yml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+global:
+  scrape_timeout: 1h
```
Comment (beorn7) [inline config/testdata/glob_timeout.good.yml]: No newline at end of file.

File: config/config_test.go [hunk 4]
```diff
@@ -89,34 +89,40 @@ var expectedConf = &Config{DefaultedConfig{
...
-		}},
+		},
 	},
-}, ""}
+	original: "",
+}
 
 func TestLoadConfig(t *testing.T) {
+	// Parse a valid file that sets a global scrape timeout.
+	if _, err := LoadFromFile("testdata/glob_timeout.good.yml"); err != nil {
```
Comment (beorn7) [inline config/config_test.go]: With the name of the file, the reader will expect a glob. But you mean "global", don't you?

Is this to test that the default conf is not overwritten by it? In that case, mention that in a comment.

Comment (fabxc) [inline config/config_test.go]: Done.

File: config/config_test.go [hunk 5]
```diff
@@ -89,34 +89,41 @@ var expectedConf = &Config{DefaultedConfig{
...
-		}},
+		},
 	},
-}, ""}
+	original: "",
+}
 
 func TestLoadConfig(t *testing.T) {
+	// Parse a valid file that sets a global scrape timeout. This tests whether parsing
+	// an overwritten default field in the global config permanentely changes the default.
```
Comment (beorn7) [inline config/config_test.go]: Typo "permanentely".

**Comment by beorn7** [comment]
Otherwise :+1:

**Pull request merged**